recipe/thirdparty/providers/facebook.go

```diff
@@ -16,110 +16,103 @@
-			accessTokenAPIParams := map[string]string{
-				"client_id":     config.ClientID,
-				"client_secret": config.ClientSecret,
-			}
-			if authCodeFromRequest != nil {
-				accessTokenAPIParams["code"] = *authCodeFromRequest
+func Facebook(input tpmodels.ProviderInput) tpmodels.TypeProvider {
+	if input.ThirdPartyID == "" {
+		input.ThirdPartyID = facebookID
+	}
+
+	if input.Config.AuthorizationEndpoint == "" {
+		input.Config.AuthorizationEndpoint = "https://www.facebook.com/v12.0/dialog/oauth"
+	}
+
+	if input.Config.TokenEndpoint == "" {
+		input.Config.TokenEndpoint = "https://graph.facebook.com/v12.0/oauth/access_token"
+	}
+
+	if input.Config.UserInfoEndpoint == "" {
+		input.Config.UserInfoEndpoint = "https://graph.facebook.com/me"
+	}
+
+	if input.Config.UserInfoMap.FromUserInfoAPI.UserId == "" {
+		input.Config.UserInfoMap.FromUserInfoAPI.UserId = "id"
+	}
+
+	if input.Config.UserInfoMap.FromUserInfoAPI.Email == "" {
+		input.Config.UserInfoMap.FromUserInfoAPI.Email = "email"
+	}
+
+	if input.Config.UserInfoMap.FromUserInfoAPI.EmailVerified == "" {
+		input.Config.UserInfoMap.FromUserInfoAPI.EmailVerified = "email_verified"
+	}
+
+	if input.Config.AuthorizationEndpointQueryParams == nil {
+		input.Config.AuthorizationEndpointQueryParams = map[string]interface{}{}
+	}
+
+	if input.Config.AuthorizationEndpointQueryParams["response_type"] == nil {
+		input.Config.AuthorizationEndpointQueryParams["response_type"] = "code"
+	}
+	if input.Config.AuthorizationEndpointQueryParams["include_granted_scopes"] == nil {
+		input.Config.AuthorizationEndpointQueryParams["include_granted_scopes"] = "true"
+	}
+	if input.Config.AuthorizationEndpointQueryParams["access_type"] == nil {
+		input.Config.AuthorizationEndpointQueryParams["access_type"] = "offline"
+	}
+
+	oOverride := input.Override
+
+	input.Override = func(provider *tpmodels.TypeProvider) *tpmodels.TypeProvider {
+		oGetConfig := provider.GetConfig
+		provider.GetConfig = func(clientType *string, input tpmodels.ProviderConfig, userContext supertokens.UserContext) (tpmodels.ProviderConfigForClient, error) {
+			config, err := oGetConfig(clientType, input, userContext)
+			if err != nil {
+				return tpmodels.ProviderConfigForClient{}, err
 			}
-			if redirectURI != nil {
-				accessTokenAPIParams["redirect_uri"] = *redirectURI
+
+			if len(config.Scope) == 0 {
+				config.Scope = []string{"email"}
 			}
 
-			authorisationRedirectURL := "https://www.facebook.com/v9.0/dialog/oauth"
-			scopes := []string{"email"}
-			if config.Scope != nil {
-				scopes = config.Scope
+			return config, err
+		}
+
+		provider.GetUserInfo = func(config tpmodels.ProviderConfigForClient, oAuthTokens tpmodels.TypeOAuthTokens, userContext supertokens.UserContext) (tpmodels.TypeUserInfo, error) {
+			queryParams := map[string]interface{}{
+				"access_token": oAuthTokens["access_token"].(string),
+				"fields":       "id,email",
+				"format":       "json",
 			}
 
```

Comment:
After this, we just call the original impl

Reply:
Done